compiler/src/module.rs

```diff
@@ -0,0 +1,295 @@
+                kind,
+            ),
+            _ => panic!("Unsupported URI scheme: {}", url.scheme()),
+        }
+    }
+    pub fn from_package_root_and_file(
+        package_root: PathBuf,
+        file: PathBuf,
+        kind: ModuleKind,
+    ) -> Self {
+        let relative_path =
+            fs::canonicalize(&file).expect("Package root does not exist or is invalid.");
+        let relative_path = match relative_path
+            .strip_prefix(fs::canonicalize(package_root.clone()).unwrap().clone())
+        {
+            Ok(path) => path,
+            Err(_) => {
+                return Module::from_anonymous_content(fs::read_to_string(file).unwrap(), kind)
+            }
+        };
+
+        let mut path = relative_path
+            .components()
+            .into_iter()
+            .map(|component| match component {
+                std::path::Component::Prefix(_) => unreachable!(),
+                std::path::Component::RootDir => unreachable!(),
+                std::path::Component::CurDir => panic!("`.` is not allowed in an module path."),
+                std::path::Component::ParentDir => {
+                    panic!("`..` is not allowed in an module path.")
+                }
+                std::path::Component::Normal(it) => {
+                    it.to_str().expect("Invalid UTF-8 in path.").to_owned()
+                }
+            })
+            .collect_vec();
+
+        if kind == ModuleKind::Code {
+            let last = path.pop().unwrap();
+            let last = last
+                .strip_suffix(".candy")
+                .expect("Code module doesn't end with `.candy`?");
+            if !last.is_empty() {
+                path.push(last.to_string());
+            }
+        }
+
+        Module {
+            package: Package::User(package_root),
+            path,
+            kind,
+        }
+    }
+}
+
+impl Package {
+    pub fn to_path(&self) -> Option<PathBuf> {
+        match self {
+            Package::User(path) => Some(path.clone()),
+            Package::External(path) => Some(path.clone()),
+            Package::Anonymous { .. } => None,
+        }
+    }
+}
+impl Module {
+    pub fn to_possible_paths(&self) -> Option<Vec<PathBuf>> {
+        let mut path = self.package.to_path()?;
+        for component in self.path.clone() {
+            path.push(component);
+        }
+        Some(match self.kind {
+            ModuleKind::Asset => vec![path],
+            ModuleKind::Code => vec![
+                {
+                    let mut path = path.clone();
+                    path.push(".candy");
+                    path
+                },
+                {
+                    let mut path = path.clone();
+                    path.set_extension("candy");
+                    path
+                },
+            ],
+        })
+    }
+
+    pub fn associated_debug_file(&self, debug_type: &str) -> PathBuf {
+        let mut path = self.to_possible_paths().unwrap().pop().unwrap();
```

Comment:
This `unwrap()` will panic when trying to save hints for an untitled file in VS Code to disk